File: RB110/small_problems/medium1/rotation_2.rb
=begin

Write a method that can rotate the last n digits of a number. For example:

Copy Code
rotate_rightmost_digits(735291, 1) == 735291
rotate_rightmost_digits(735291, 2) == 735219
rotate_rightmost_digits(735291, 3) == 735912
rotate_rightmost_digits(735291, 4) == 732915
rotate_rightmost_digits(735291, 5) == 752913
rotate_rightmost_digits(735291, 6) == 352917
Note that rotating just 1 digit results in the original number being returned.

You may use the rotate_array method from the previous exercise if you want. (Recommended!)

You may assume that n is always a positive integer.

Input:
  -Integer of multiple digits
  -a single integer representing the amount of digits to be rotated, ststrting from the end

Output:
  -Integer of multiple digits with the digits rotated by the input integer

Rules
  -Explicit
    -Rotating digits by just one should return the same digits
  -Implicit:
    -Return an integer

Questions?
  -Are we allowed to use the rotate method?

Test cases // Exsamples

rotate_rightmost_digits(735291, 1) == 735291
rotate_rightmost_digits(735291, 2) == 735219
rotate_rightmost_digits(735291, 3) == 735912
rotate_rightmost_digits(735291, 4) == 732915
rotate_rightmost_digits(735291, 5) == 752913
rotate_rightmost_digits(735291, 6) == 352917

data structre:
  - Integers
assisting structure:
  - Array

Algorithm
  -If rotating number is equal or less that 1 return digits
  -Take number and turn into array
  -take rotating number and make it negative * 1 save to index
  -Create an array starting from 0 to the provided index
  -Rotate the remaining numbers 
  -Join the array, covert to ingteger and return.

=end

def rotate_array(arr)
  arr.values_at(1..(arr.size - 1), 0)
end

def rotate_rightmost_digits(digits, num)
  rotate_array = digits.to_s.chars
  index = num * -1

  first_half = rotate_array.values_at(0..(index - 1))
  second_half = rotate_array.values_at(index..-1)

  rotated_digits = first_half + rotate_array(second_half)

  rotated_digits.join.to_i
end


p rotate_rightmost_digits(735291, 1) == 735291
p rotate_rightmost_digits(735291, 2) == 735219
p rotate_rightmost_digits(735291, 3) == 735912
p rotate_rightmost_digits(735291, 4) == 732915
p rotate_rightmost_digits(735291, 5) == 752913
p rotate_rightmost_digits(735291, 6) == 352917

File: LS215/exerises/medium1/rotation2.js
/*
Write a function that rotates the last n digits of a number. For the rotation, rotate by one digit to the 
left, moving the first digit to the end.

Input:
  - A number or Digits
  - 'n' the number of digits from the right moving left rotate the digits, moving the first digits to the end
Output:
  - Number

Rules:
  - Return a number
  - Rotate the numbers starting from the left 'n' numbers
  - Leave the first part of the number intact
  - 

Questions:
  - Will there always be 2 arguments
  - Will the first argument always be a number? Will the second argument always be a number
  - Will the second argument ever be 0?

Data Structure:
  - Convert the number to strings
  - Convert the strings into an array
  - Return a number

Algorithm:
  - Convert the input numbers into strings then split into an array
  - Get the elements that need to be rotated
    - toRotate = array[array.length - n]
  - Get the first elements in the array 
    - array.slice(0, array length - n) 
  - Call the rotateArray function on toRotate
  - Concat the first elements and the rotated elements
  - join the array together and covert back to numbers
  - Return numbers

*/

function rotateArray(input) {
  if (!Array.isArray(input)) return undefined;
  if (input.length === 0) return [];
  if (input === undefined) return undefined;

  let copyOfArray = input.slice();
  let firstEle = copyOfArray.shift();
  copyOfArray[copyOfArray.length] = firstEle;

  return copyOfArray;
}

function rotateRightmostDigits(digits, n) {
  let array = digits.toString().split('');
  let toRotate = array.slice((array.length - n), array.length);
  let first = array.slice(0, array.length - n);

  let rotated = first.concat(rotateArray(toRotate));

  return Number(rotated.join(''));
}

console.log(rotateRightmostDigits(735291, 1));      // 735291
console.log(rotateRightmostDigits(735291, 2));      // 735219
console.log(rotateRightmostDigits(735291, 3));      // 735912
console.log(rotateRightmostDigits(735291, 4));      // 732915
console.log(rotateRightmostDigits(735291, 5));      // 752913
console.log(rotateRightmostDigits(735291, 6));      // 352917